## Connecting to the Database

In [2]:
pip install psycopg2-binary;

  Using cached psycopg2_binary-2.8.6-cp37-cp37m-manylinux1_x86_64.whl (3.0 MB)
Note: you may need to restart the kernel to use updated packages.


In [3]:
pip install mxnet

  Using cached mxnet-1.7.0.post2-py2.py3-none-manylinux2014_x86_64.whl (54.7 MB)
  Using cached graphviz-0.8.4-py2.py3-none-any.whl (16 kB)
Note: you may need to restart the kernel to use updated packages.


In [17]:
import psycopg2

In [18]:
# Put the password in 
PGHOST = 'tripdatabase2.cmaaautpgbsf.us-east-2.rds.amazonaws.com'
PGDATABASE = ''
PGUSER = 'postgres'
PGPASSWORD = 'Josh1234'

In [19]:
# Database Context Manager
try:   
    # Set up a connection to the postgres server.    
    conn = psycopg2.connect(user = PGUSER,
                            port = "5432",
                            password = PGPASSWORD,
                            host = PGHOST,
                            database = PGDATABASE)
    # Create a cursor object
    cursor = conn.cursor()   
    cursor.execute("SELECT version();")
    record = cursor.fetchone()
    print("Connection Success:", record,"\n")

except (Exception, psycopg2.Error) as error:
    print("Error while connecting to PostgreSQL", error)

Connection Success: ('PostgreSQL 12.5 on x86_64-pc-linux-gnu, compiled by gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-11), 64-bit',) 



In [20]:
import pandas as pd
import numpy as np
import Queries
import matplotlib.pyplot as plt
import seaborn as sns
import importlib
import s3fs
importlib.reload(Queries)

<module 'Queries' from '/root/Citi-Bike-Expansion/Queries.py'>

In [21]:
# The S3 Bucket that will be used to store the data should be created beforehand
ACCESS_KEY_ID = 'AKIARJEUISD2VILSZ6HM'
ACCESS_SECRET_KEY = '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'

fs = s3fs.S3FileSystem(anon=False, key = ACCESS_KEY_ID, secret= ACCESS_SECRET_KEY)

The key data file that is needed to do the machine learning contains data for every zipcode in the United States. The aim is to come up with a universal classification model that can be used for any location in the States. Therefore, a couple of columns are going to be dropped from the start:
- Country Name 
- State Name & State Code

Some other columns that will be dropped are:
- Cities Postal Code: These codes are not the same as the zipcodes that we have).
- Area in Miles: We will continue to work in meters.
- Total Housing: There are two identical columns, we drop one.
- Predicted Qualified Leads: Not actually a zipcode property.

In [22]:
zipcode_file = "s3://williams-citibike/zipcodes.csv"

with fs.open(zipcode_file, 'r') as file:
    zipcodes = pd.read_csv(file, sep=',', low_memory = False)
    
zipcodes = zipcodes[zipcodes.country_name == 'United States']
zipcodes.drop(columns = ['cities_postalcode_id', 'state_name', 'state_code', 
                         'country_name','area_land_sq_miles', 'area_water_sq_miles',
                         'units_in_structure_housing_units_total_housing_units',
                         'predicted_qualified_leads_per_month'], inplace=True)

In [60]:
zipcodes.tail()

zip_code                      cbsa_type  \
43568    99926                            NaN   
43569    99927                            NaN   
43570    99928  Micropolitan Statistical Area   
43571    99929                            NaN   
43572    99950                            NaN   

                                   zip_type  area_land_sq_meters  \
43568                         Zip Code Area         3.439446e+08   
43569                         Zip Code Area         5.896901e+08   
43570  Post Office or large volume customer         1.242480e+10   
43571                         Zip Code Area         5.635591e+09   
43572                         Zip Code Area         5.285066e+09   

       area_water_sq_meters  sex_and_age_population_total_population  \
43568          2.928365e+08                                   1667.0   
43569          1.800157e+07                                     81.0   
43570          1.548638e+09                                  13745.0   
43571          6.377141e+08                                   2385.0   
43572          1.132247e+09                                   1800.0   

       sex_and_age_population_18_years_and_over_total_population  \
43568                                             1163.0           
43569                                               81.0           
43570                                            10644.0           
43571                                             1890.0           
43572                                             1407.0           

       sex_and_age_age_median_age_years_total_population  \
43568                                               32.4   
43569                                               49.9   
43570                                               39.2   
43571                                               48.6   
43572                                               41.5   

       employment_status_percent_in_labor_force_of_population_16_years_and_over  \
43568                                              0.594                          
43569                                              0.605                          
43570                                              0.694                          
43571                                              0.604                          
43572                                              0.698                          

       employment_status_percent_unemployed_of_civilian_labor_force_population_16_years_and_over  \
43568                                              0.143                                           
43569                                              0.000                                           
43570                                              0.075                                           
43571                                              0.072                                           
43572                                              0.088                                           

       ...  \
43568  ...   
43569  ...   
43570  ...   
43571  ...   
43572  ...   

       units_in_structure_percent_20_or_more_units_of_total_housing_units  \
43568                                              0.023                    
43569                                              0.000                    
43570                                              0.098                    
43571                                              0.033                    
43572                                              0.000                    

       households_by_type_population_average_household_size  \
43568                                               3.70      
43569                                               1.40      
43570                                               2.58      
43571                                               2.25      
43572                                               2.55      

       educational_attainment_percent_bachelors_degree_or_higher_of_popu

In [24]:
station_zipcodes = Queries.get_zipcode_stations(conn)
station_zipcodes.zipcode = station_zipcodes.astype('str')
station_zipcodes.zipcode = station_zipcodes.zipcode.str.zfill(5)

In [25]:
station_zipcodes.head()

zipcode  num_stations
0   000-1             1
1   02108             4
2   02109             2
3   02110            10
4   02111             5

In [26]:
bikeshare_zipcodes = zipcodes.merge(station_zipcodes, how='inner', left_on='zip_code', right_on='zipcode').drop(columns=['zipcode'])

In [31]:
column_renames = ['zipcode', 'cbsa', 'ziptype', 'land_area_sqm',
                  'water_area_sqm', 'total_population', 'total_population_18_over',
                  'median_age', 'pct_labor_force_16_over', 
                  'pct_labor_force_unemployed_16_over', 'pct_armed_force_16_over', 
                  'pct_labor_force_employed_16_over', 'median_household_income',
                  'family_income_per_capita', 'total_housing_units',
                  'pct_vacant_housing', 'occupied_housing_units',
                  'pct_no_vehicle_of_occupied_housing', 'median_price_owner_occupied_units',
                  'median_rent_occupied_units_paying_rent', 'pct_1_unit_attached',
                  'pct_1_unit_detached', 'pct_2_units', 'pct_3-4_units', 'pct_5-9_units',
                  'pct_10-19_units', 'pct_20+_units', 'avg_household_size', 'pct_bachelors+_25_over',
                  'pct_diff_housing_from_last_year', 'pct_same_housing_from_last_year',
                  'resident_since_last_year', 'population_in_college_grad', 'population_density', 
                  'pct_population_in_college_grad', 'zipcode_segment', 'num_stations'
                 ]

In [32]:
bikeshare_zipcodes.columns = column_renames

In [40]:
bikeshare_zipcodes.head()

zipcode                           cbsa        ziptype  land_area_sqm  \
0   02108  Metropolitan Statistical Area  Zip Code Area       353039.0   
1   02109  Metropolitan Statistical Area  Zip Code Area       451168.0   
2   02110  Metropolitan Statistical Area  Zip Code Area       477072.0   
3   02111  Metropolitan Statistical Area  Zip Code Area       699955.0   
4   02113  Metropolitan Statistical Area  Zip Code Area       259756.0   

   water_area_sqm  total_population  total_population_18_over  median_age  \
0             0.0            4049.0                    3622.0        35.2   
1        292691.0            4015.0                    3713.0        32.9   
2        234496.0            2124.0                    2012.0        53.6   
3         43757.0            7564.0                    6790.0        32.9   
4             0.0            7495.0                    7188.0        30.2   

   pct_labor_force_16_over  pct_labor_force_unemployed_16_over  ...  \
0                    0.699                               0.077  ...   
1                    0.749                               0.053  ...   
2                    0.670                               0.018  ...   
3                    0.624                               0.088  ...   
4                    0.824                               0.020  ...   

   avg_household_size  pct_bachelors+_25_over  \
0                1.70                   0.829   
1                1.72                   0.736   
2                1.61                   0.874   
3                2.00                   0.526   
4                1.72                   0.772   

   pct_diff_housing_from_last_year  pct_same_housing_from_last_year  \
0                            0.263                            0.728   
1                            0.324                            0.625   
2                            0.176                            0.806   
3                            0.203                            0.764   
4                            0.309                            0.680   

   resident_since_last_year  population_in_college_grad  population_density  \
0                    4049.0                       580.0        29772.058824   
1                    3978.0                       235.0        23074.712644   
2                    2124.0                       197.0        11543.478261   
3                    7493.0                      2069.0        28014.814815   
4                    7495.0                       976.0        74950.000000   

   pct_population_in_college_grad                  zipcode_segment  \
0                        0.143245     Medium density professionals   
1                        0.058531     Medium density professionals   
2                        0.092750        Low density professionals   
3                        0.273533                 College students   
4                        0.130220  Very high density professionals   

   num_stations  
0             4  
1             2  
2            10  
3             5  
4             2  

[5 rows x 37 columns]

CBSA Type (Categorical): 
There are two core based statistical area (cbsa) types in the United States: 
- Metropolitan: one urbanized area of 50,000 or more inhabitants.
- Micropolitan: one urban cluster of at least 10,000 but less than 50,000 population. 

***Observation: All the zipcodes that have bikeshare stations are metropolitan statistical areas. (See Commented Code Below)***
https://www.census.gov/programs-surveys/metro-micro/about.html#:~:text=Each%20metropolitan%20statistical%20area%20must,but%20less%20than%2050%2C000%20population
https://en.wikipedia.org/wiki/Core-based_statistical_area


In [ ]:
# bikeshare_zipcodes.cbsa.unique()

Zip Type (Categorical): There are two types of zip codes
- Zip Code Areas are typical ZIP Codes without any special designation. 
- Post Office or large volume customers typically occupy a very small geography and have few residential addresses located within them. 

https://udsmapper.org/zip-code-to-zcta-crosswalk/

Zip Code Segment (Categorical):
Describes the density and the type of people that are working in the area

In [38]:
bikeshare_zipcodes.shape

(318, 37)

Let's start by dealing with missing data.

In [50]:
bikeshare_zipcodes.isna().sum()

zipcode                                    0
cbsa                                       1
ziptype                                    0
land_area_sqm                              0
water_area_sqm                             0
total_population                           0
total_population_18_over                   0
median_age                                13
pct_labor_force_16_over                    0
pct_labor_force_unemployed_16_over         0
pct_armed_force_16_over                    0
pct_labor_force_employed_16_over           0
median_household_income                   16
family_income_per_capita                  13
total_housing_units                        0
pct_vacant_housing                         0
occupied_housing_units                     0
pct_no_vehicle_of_occupied_housing         0
median_price_owner_occupied_units         18
median_rent_occupied_units_paying_rent    15
pct_1_unit_attached                        0
pct_1_unit_detached                        0
pct_2_unit

**Median Age**

In [74]:
bikeshare_zipcodes[bikeshare_zipcodes.median_age.isna()].head()

Empty DataFrame
Columns: [zipcode, cbsa, ziptype, land_area_sqm, water_area_sqm, total_population, total_population_18_over, median_age, pct_labor_force_16_over, pct_labor_force_unemployed_16_over, pct_armed_force_16_over, pct_labor_force_employed_16_over, median_household_income, family_income_per_capita, total_housing_units, pct_vacant_housing, occupied_housing_units, pct_no_vehicle_of_occupied_housing, median_price_owner_occupied_units, median_rent_occupied_units_paying_rent, pct_1_unit_attached, pct_1_unit_detached, pct_2_units, pct_3-4_units, pct_5-9_units, pct_10-19_units, pct_20+_units, avg_household_size, pct_bachelors+_25_over, pct_diff_housing_from_last_year, pct_same_housing_from_last_year, resident_since_last_year, population_in_college_grad, population_density, pct_population_in_college_grad, zipcode_segment, num_stations]
Index: []

[0 rows x 37 columns]

Zipcodes that have missing data for the median has missing data for all the columns. Unfortunately, this data doesn't exist anywhere.

In [68]:
bikeshare_zipcodes = bikeshare_zipcodes.drop(bikeshare_zipcodes[bikeshare_zipcodes.median_age.isna()].index)

In [69]:
bikeshare_zipcodes.isna().sum()

zipcode                                    0
cbsa                                       1
ziptype                                    0
land_area_sqm                              0
water_area_sqm                             0
total_population                           0
total_population_18_over                   0
median_age                                 0
pct_labor_force_16_over                    0
pct_labor_force_unemployed_16_over         0
pct_armed_force_16_over                    0
pct_labor_force_employed_16_over           0
median_household_income                    3
family_income_per_capita                   0
total_housing_units                        0
pct_vacant_housing                         0
occupied_housing_units                     0
pct_no_vehicle_of_occupied_housing         0
median_price_owner_occupied_units          5
median_rent_occupied_units_paying_rent     2
pct_1_unit_attached                        0
pct_1_unit_detached                        0
pct_2_unit

**Median Household Income**

In [70]:
bikeshare_zipcodes[bikeshare_zipcodes.median_household_income.isna()]

zipcode                           cbsa        ziptype  land_area_sqm  \
94    10282  Metropolitan Statistical Area  Zip Code Area       179389.0   
152   20052  Metropolitan Statistical Area  Zip Code Area         7062.0   
309   94720  Metropolitan Statistical Area  Zip Code Area      1610594.0   

     water_area_sqm  total_population  total_population_18_over  median_age  \
94          55392.0            5635.0                    4036.0        34.5   
152             0.0             431.0                     431.0        19.8   
309             0.0            3166.0                    3048.0        19.7   

     pct_labor_force_16_over  pct_labor_force_unemployed_16_over  ...  \
94                     0.781                               0.022  ...   
152                    0.357                               0.201  ...   
309                    0.307                               0.233  ...   

     avg_household_size  pct_bachelors+_25_over  \
94                 2.32                   0.857   
152                 NaN                   0.852   
309                3.17                   0.889   

     pct_diff_housing_from_last_year  pct_same_housing_from_last_year  \
94                             0.171                            0.785   
152                            0.619                            0.281   
309                            0.609                            0.211   

     resident_since_last_year  population_in_college_grad  population_density  \
94                     5419.0                       165.0        81666.666667   
152                     431.0                       423.0       143666.666667   
309                    3166.0                      3094.0         5090.032154   

     pct_population_in_college_grad             zipcode_segment  num_stations  
94                         0.029281  High density professionals             2  
152                        0.981439            College students             2  
309                        0.977258            College students             2  

[3 rows x 37 columns]

Median Household Income for **10282** from 1st source 250,000 and was confirmed by 2nd Source. No data for **20052** in any source. No Median Income but, there is Average Household Income for **94720** in 2nd source at 319,462.

First Source
https://www.census.gov/acs/www/data/data-tables-and-tools/narrative-profiles/2019/

Second Check
https://www.incomebyzipcode.com/

In [73]:
bikeshare_zipcodes.loc[94,'median_household_income'] = 250000
bikeshare_zipcodes.loc[309, 'median_household_income'] = 319462

**Median Price of Owner Occupied Housing**

In [75]:
bikeshare_zipcodes[bikeshare_zipcodes.median_price_owner_occupied_units.isna()]

zipcode                           cbsa        ziptype  land_area_sqm  \
34    02163  Metropolitan Statistical Area  Zip Code Area       256728.0   
36    02199  Metropolitan Statistical Area  Zip Code Area       148225.0   
60    10013  Metropolitan Statistical Area  Zip Code Area      1425085.0   
152   20052  Metropolitan Statistical Area  Zip Code Area         7062.0   
309   94720  Metropolitan Statistical Area  Zip Code Area      1610594.0   

     water_area_sqm  total_population  total_population_18_over  median_age  \
34          32166.0            2575.0                    2451.0        27.4   
36              0.0            1260.0                    1218.0        53.4   
60              0.0           28211.0                   23643.0        36.9   
152             0.0             431.0                     431.0        19.8   
309             0.0            3166.0                    3048.0        19.7   

     pct_labor_force_16_over  pct_labor_force_unemployed_16_over  ...  \
34                     0.242                               0.100  ...   
36                     0.580                               0.048  ...   
60                     0.672                               0.044  ...   
152                    0.357                               0.201  ...   
309                    0.307                               0.233  ...   

     avg_household_size  pct_bachelors+_25_over  \
34                 2.02                   0.980   
36                 1.55                   0.868   
60                 2.20                   0.627   
152                 NaN                   0.852   
309                3.17                   0.889   

     pct_diff_housing_from_last_year  pct_same_housing_from_last_year  \
34                             0.391                            0.468   
36                             0.357                            0.628   
60                             0.134                            0.844   
152                            0.619                            0.281   
309                            0.609                            0.211   

     resident_since_last_year  population_in_college_grad  population_density  \
34                     2562.0                      2150.0        26010.101010   
36                     1243.0                        71.0        22105.263158   
60                    27842.0                      2507.0        51292.727273   
152                     431.0                       423.0       143666.666667   
309                    3166.0                      3094.0         5090.032154   

     pct_population_in_college_grad               zipcode_segment  \
34                         0.834951              College students   
36                         0.056349  Medium density professionals   
60                         0.088866    High density professionals   
152                        0.981439              College students   
309                        0.977258              College students   

     num_stations  
34              2  
36              2  
60             21  
152             2  
309             2  

[5 rows x 37 columns]

For both the NYC 10013 and the California 94720 zipcode the median owner occupied price was more than 2,000,000

In [76]:
bikeshare_zipcodes.loc[60,'median_price_owner_occupied_units'] = 2000000
bikeshare_zipcodes.loc[309,'median_price_owner_occupied_units'] = 2000000